common/djangoapps/third_party_auth/utils.py

```diff
@@ -172,16 +172,25 @@ def get_user_from_email(details):
 
 def create_or_update_bulk_saml_provider_data(entity_id, public_keys, sso_url, expires_at):
     """
-    Placeholder
+    Method to bulk update or create provider data entries
     """
     fetched_at = now()
     new_records_created = False
     # Create a data record for each of the public keys provided
     for key in public_keys:
-        _, created = SAMLProviderData.objects.update_or_create(
-            public_key=key, entity_id=entity_id,
-            defaults={'sso_url': sso_url, 'expires_at': expires_at, 'fetched_at': fetched_at},
-        )
+        existing_data_objects = SAMLProviderData.objects.filter(public_key=key, entity_id=entity_id)
+        if len(existing_data_objects) > 1:
```

Comment:
So this is basically saying "If there is more than one SAMLProviderData record having this key+entity_id pair, update all of the records to have the same sso_url and expire/fetched timestamps." right?  

I don't think SAMLProviderData is a ConfigurationModel, so history should not be an issue here -- thus, if we have multiple records with same key+entity_id pair, it's an exceptional condition and something that probably should be cleaned up.  Unless I'm thinking about this incorrectly?

Reply:
The more likely situation is that we have multiple SAMLProvider records for a given entity_id, where the key value is different.  The question to ask in this case is "When do we consider one of these records stale/irrelevant?".  

I'm not sure if we can answer that without decoding the cert and scooping up the actual expiration date attribute value -- which I think we should be doing vs. setting an arbitrary expiration date.

Reply:
This was/is mostly to account for saml test ID's on stage. Because saml testing asks for and supplies identical keys/certs/redirects every usage, it was resulting in many identical data configs being generated on stage.